Sources/GRPC/ConnectionPool/PoolManager.swift

```diff
@@ -0,0 +1,270 @@
+    // practice this is unlikely to happen unless a custom `EventLoopGroup` is constructed, because
+    // of that we'll only check when running in debug mode.
+    debugOnly {
+      let eventLoopIDs = group.makeIterator().map { ObjectIdentifier($0) }
+      let uniqueEventLoopIDs = Set(eventLoopIDs)
+      assert(
+        eventLoopIDs.count == uniqueEventLoopIDs.count,
+        "'group' contains non-unique event loops"
+      )
+    }
+  }
+
+  deinit {
+    self.lock.withLockVoid {
+      assert(
+        self.state.isShutdownOrShuttingDown,
+        "The pool manager (\(ObjectIdentifier(self))) must be shutdown before going out of scope."
+      )
+    }
+  }
+
+  /// Initialize the pool manager, create and initialize one connection pool per event loop in the
+  /// pools `EventLoopGroup`.
+  ///
+  /// - Important: Must only be called once.
+  /// - Parameters:
+  ///   - configuration: The configuration used for each connection pool.
+  ///   - logger: A logger.
+  private func initialize(
+    perPoolConfiguration configuration: PerPoolConfiguration,
+    logger: GRPCLogger
+  ) {
+    var logger = logger
+    logger[metadataKey: Metadata.id] = "\(ObjectIdentifier(self))"
+
+    let pools = self.makePools(perPoolConfiguration: configuration, logger: logger)
+
+    logger.debug("initializing connection pool manager", metadata: [
+      Metadata.poolCount: "\(pools.count)",
+      Metadata.connectionsPerPool: "\(configuration.connections)",
+      Metadata.waitersPerPool: "\(configuration.maximumWaiters)",
+    ])
+
+    // The assumed maximum number of streams concurrently available in each pool.
+    let assumedCapacity = configuration.assumedStreamCapacity
+
+    self.lock.withLockVoid {
+      // We'll blow up if we've already been initialized, that's fine, we don't allow callers to
+      // call `initialize` directly.
+      self.state.activate(pools: pools, assumingPerPoolCapacity: assumedCapacity)
+    }
+
+    for pool in pools {
+      pool.initialize(connections: configuration.connections)
+    }
+  }
+
+  /// Make one pool per `EventLoop` in the pool's `EventLoopGroup`.
+  /// - Parameters:
+  ///   - configuration: The configuration to make each pool with.
+  ///   - logger: A logger.
+  /// - Returns: An array of `ConnectionPool`s.
+  private func makePools(
+    perPoolConfiguration configuration: PerPoolConfiguration,
+    logger: GRPCLogger
+  ) -> [ConnectionPool] {
+    let eventLoops = self.group.makeIterator()
+    return eventLoops.map { eventLoop in
+      // We're creating a retain cycle here as each pool will reference the manager and the per-pool
+      // state will hold the pool which will in turn be held by the pool manager. That's okay: when
+      // the pool is shutdown the per-pool state and in turn each connection pool will be dropped.
+      // and we'll break the cycle.
+      return ConnectionPool(
+        eventLoop: eventLoop, maxWaiters: configuration.maximumWaiters,
```

Comment:
Nit: I don't love having `maxWaiters` stashed over here to the side. Can we give it its own line?